codegen/render/kube_multicluster_test.go

```diff
@@ -0,0 +1,403 @@
+					return paintDeletes.get(multicluster.MasterCluster, paint2.Name)
+				}, time.Second).Should(Equal(reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      paint2.Name,
+					Namespace: paint2.Namespace,
+				}}))
+
+				// Sleep the amount of time we typically allow for a reconcile to happen.
+				time.Sleep(1 * time.Second)
+				Expect(paintDeletes.get(cluster2, paint2.Name)).To(Equal(reconcile.Request{}))
+
+				/**
+				When a KubeConfig secret is restored, paint is again reconciled for that cluster
+				*/
+
+				_, err = kubehelp.MustKubeClient().CoreV1().Secrets(ns).Create(mustNewKubeConfigSecret())
+				Expect(err).NotTo(HaveOccurred())
+
+				paint3 := newPaint(ns, "paint-mc-reconciler-3")
+				err = clientSet.Paints().CreatePaint(ctx, paint3)
+				Expect(err).NotTo(HaveOccurred())
+
+				mustReconcile(paint3, paintMap, paintDeletes)
+			})
+
+			It("works when a loop is registered after the watcher is started", func() {
+				err := cw.Run(masterManager)
+				Expect(err).NotTo(HaveOccurred())
+
+				loop := controller.NewMulticlusterPaintReconcileLoop("mid-run-paint", cw)
+
+				midRunReconciledPaint := newConcurrentPaintMap()
+				midRunReconciledDeleteRequests := newConcurrentRequestMap()
+
+				loop.AddMulticlusterPaintReconciler(ctx, &controller.MulticlusterPaintReconcilerFuncs{
+					OnReconcilePaint: func(clusterName string, obj *Paint) (result reconcile.Result, e error) {
+						midRunReconciledPaint.add(clusterName, obj)
+						return result, e
+					},
+					OnReconcilePaintDeletion: func(clusterName string, req reconcile.Request) {
+						midRunReconciledDeleteRequests.add(clusterName, req)
+					},
+				})
+
+				Eventually(func() *Paint {
+					return midRunReconciledPaint.get(multicluster.MasterCluster, paint.Name)
+				}, time.Second).ShouldNot(BeNil())
+
+				Eventually(func() *Paint {
+					return midRunReconciledPaint.get(cluster2, paint.Name)
+				}, time.Second).ShouldNot(BeNil())
+
+				err = clientSet.Paints().DeletePaint(ctx, client.ObjectKey{
+					Name:      paint.Name,
+					Namespace: paint.Namespace,
+				})
+				Expect(err).NotTo(HaveOccurred())
+
+				Eventually(func() reconcile.Request {
+					return midRunReconciledDeleteRequests.get(multicluster.MasterCluster, paint.Name)
+				}, time.Second).Should(Equal(reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      paint.Name,
+					Namespace: paint.Namespace,
+				}}))
+
+				Eventually(func() reconcile.Request {
+					return midRunReconciledDeleteRequests.get(cluster2, paint.Name)
+				}, time.Second).Should(Equal(reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      paint.Name,
+					Namespace: paint.Namespace,
+				}}))
+			})
+		})
+	})
+})
+
+type concurrentPaintMap struct {
+	m     map[string]*Paint
+	mutex sync.Mutex
+}
+
+func newConcurrentPaintMap() *concurrentPaintMap {
+	return &concurrentPaintMap{
+		m:     make(map[string]*Paint),
+		mutex: sync.Mutex{},
+	}
+}
```

Comment:
Since this is test code anyway, maybe we should use https://golang.org/src/sync/map.go rather than creating all this boilerplate

Reply:
good tip! ended up using this to simplify the boilerplatey concurrent maps implemented here, as their API is a bit more readable in the context of the tests imo